### DISCUSSION TITLE
Does pocketbase handles concurrent read-write efficiently?

### DISCUSSION THREAD

**Discussion by CovertCode** [pocketbase/pocketbase#4209] [Q&A] [answered]
Is there page where I can know how pocketbase handles concurrent read-write and db lock during write process?
Because traditionally read is blocked while there's a write process on going. So it would be better for me to know about this while using pocketbase to avoid issues during heavy reads.

**Comment by ganigeorgiev** [comment]
> Because traditionally read is blocked while there's a write process on going

That is not true for the majority of the queries PocketBase execute.
We use SQLite in WAL mode which allows multiple readers and a single writer (writes are queue up) to operate concurrently. For more details you can check the official SQLite documentation - https://www.sqlite.org/wal.html.

The only exception to the above is when you have mixed transactions, but those cases are limited (eg. on record delete) and gracefully handled behind the scene with the help of the SQLite busy_timeout and the additional auto fail/retry on server level.

**Reply by CovertCode** [reply]
Is there any fixed number of multiple readers in pocketbase or it's dynamic and can be changed by user during compile?
Lets say if there are 100 write request and 20k read request simultaneously, then what will happen in this scenario?

**Reply by ganigeorgiev** [reply, marked as answer]
There is a limit to the max number of open connections to avoid excessive memory usage, but keep in mind that connections are reused.
We have 2 separate pools for reads and write connections. For reads the default is 120, for writes we allow only 1 open connection at a time because you can't have more than 1 writer (_you can change the default connections limit via [`pocketbase.NewWithConfig(config)`](https://pkg.go.dev/github.com/pocketbase/pocketbase#NewWithConfig)_).
In your specific example, writes will queue up (aka. they will execute one by one) while the reads will fire up to 120 connections and will try to execute the queries concurrently (_this of course will also depends on how many cores your program is allowed to use_). If you are looking for benchmarks you can explore the [benchmarks repo](https://github.com/pocketbase/benchmarks).
